File: week16/유리/BJ3079_입국심사.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
	static long n, m, result = Long.MAX_VALUE;
	static long [] data;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer str = new StringTokenizer(br.readLine());
		
		n = Long.parseLong(str.nextToken());
		m = Long.parseLong(str.nextToken());
		long max = 0;
		data = new long[(int)n];
		for(int i = 0; i < n; i++) {
			data[i] = Long.parseLong(br.readLine());
			max = Math.max(max, data[i]);
		}
		
		Arrays.sort(data);
		
		long left = 0;
		long right = max*m;
		while(left <= right) {
			long mid = (left+right)/2;
			
			long cnt = 0;
			for(long tmp : data) {
				if(cnt >= m) {
					break;
				}
				cnt +=mid/tmp;
			}
			
			if(cnt<m) {
				left= mid+1;
			}else {
				right=mid-1;
				result = Math.min(result, mid);
			}
		}
		
		System.out.println(result);
	}
}

File: week16/보연/BJ3079_입국심사.cpp
#include <iostream>
#include <algorithm>

using namespace std;

typedef long long ll;

const int SIZE = 100001;
int N;
ll M;
int t[SIZE];

void input() {
    cin >> N >> M;
    for(int i=0; i<N; i++) {
        cin >> t[i];
    }
}

void solve() {
    sort(t, t + N);
    
    ll left = 1;
    ll right = t[N-1] * M;
    
    while(left <= right) {
        ll mid = (left + right) / 2;
        ll cnt = 0;
        for(int i=0; i<N; i++) {
            cnt += (mid / t[i]);
            if(cnt > M) break; //이미 초과했으므로 가지치기
        }
        
        //cout << left << " " << right << "=>" << cnt <<'\n';
        
        if(cnt >= M) {
            //가능한 수가 많음 -> 시간을 더 줄이기
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    cout << left << '\n';
    
}

int main(){
    input();
    solve();
    return 0;
}
